lib/active_merchant/billing/gateways/payu_latam.rb

```diff
@@ -152,28 +153,60 @@ def add_order(post, options)
+        payer[:birthdate] = options[:birth_date] if options[:birth_date] && options[:payment_country] == 'MX'
+        payer[:billingAddress] = add_billing_address(post, options)
+        post[:transaction][:payer] = payer
+      end
+
+      def add_billing_address(post, options)
+        address = options[:billing_address]
+        billing_address = {}
+        billing_address[:street1] = address[:address1]
+        billing_address[:street2] = address[:address2]
+        billing_address[:city] = address[:city]
+        billing_address[:state] = address[:state]
+        billing_address[:country] = address[:country]
+        billing_address[:postalCode] = address[:zip] if options[:payment_country] == 'MX'
+        billing_address[:phone] = address[:phone]
+        billing_address
+      end
+
       def add_buyer(post, options)
-        if address = options[:shipping_address]
-          buyer = {}
-          buyer[:fullName] = address[:name]
-          buyer[:dniNumber] = options[:dni_number] if options[:dni_number]
-          buyer[:dniType] = options[:dni_type] if options[:dni_type]
-          buyer[:emailAddress] = options[:email] if options[:email]
-          buyer[:contactPhone] = address[:phone]
-          shipping_address = {}
-          shipping_address[:street1] = address[:address1]
-          shipping_address[:street2] = address[:address2]
-          shipping_address[:city] = address[:city]
-          shipping_address[:state] = address[:state]
-          shipping_address[:country] = address[:country]
-          shipping_address[:postalCode] = address[:zip]
-          shipping_address[:phone] = address[:phone]
-          buyer[:shippingAddress] = shipping_address
-          post[:transaction][:order][:buyer] = buyer
-        end
+        buyer = {}
+        buyer[:fullName] = options[:buyer_name] if options[:buyer_name]
+        buyer[:dniNumber] = options[:buyer_dni_number] if options[:buyer_dni_number]
+        buyer[:dniType] = options[:buyer_dni_type] if options[:buyer_dni_type]
+        buyer[:cnpj] = options[:buyer_cnpj] if options[:buyer_cnpj] && options[:payment_country] == 'BR'
+        buyer[:emailAddress] = options[:buyer_email] if options[:buyer_email]
+        buyer[:contactPhone] = options[:shipping_address][:phone] if (options[:shipping_address] && options[:shipping_address][:phone])
+        buyer[:shippingAddress] = add_shipping_address(post, options) if options[:shipping_address]
+        post[:transaction][:order][:buyer] = buyer
+      end
+
+      def add_shipping_address(post, options)
+        address = options[:shipping_address]
+        shipping_address = {}
+        shipping_address[:street1] = address[:address1]
+        shipping_address[:street2] = address[:address2]
+        shipping_address[:city] = address[:city]
+        shipping_address[:state] = address[:state]
+        shipping_address[:country] = address[:country]
+        shipping_address[:postalCode] = address[:zip]
```

Comment:
Is the `if options[:payment_country] == 'MX'` qualifier for `:postalCode` only needed for the billing address and not the shipping address?

Reply:
Yep. The shipping address could be somewhere else, so I think it's not validated.